Implement ARC KV cache eviction policy

albertoperdomo2 opened the pull request:
## Purpose

Current vLLM v1 uses LRU for KV cache eviction, which works well for scan-heavy workloads but can struggle with mixed access patterns. ARC automatically adapts between recency-based (LRU-like) and frequency-based (LFU-like) eviction, providing better cache efficiency for diverse workloads.

### Algorithm logic
This is my implementation of the famous [Adaptive Replacement Cache](https://en.wikipedia.org/wiki/Adaptive_replacement_cache) eviction policy but in this case for KV cache. 

The core data structures are:
- `T1` is the recent cache i.e. blocks accessed once. 
- `T2` is the frequent cache i.e. blocks accessed multiple times. 
- `B1/B2` are ghost lists for `T1/T2` , tracking recently evicted blocks. 
- `target_t1_size` is an adaptive target size for T1 partition. 

So, in terms of how the **algorithm flow**:

1. Cache lookup (`lookup`): Search T1 and T2 for block hashes and counts consecutive hits until miss or non-ready block. 
2. Cache touch (`touch`): This is the “adaptative learning” part. For each block_hash (reverse order) we check:
   - If it's in T1, move it to T2 (promotion from recent to frequent). 
   - If it’s in T2, move to the end (MRU position).
   - If it’s in B1 ghost list, increase target_t1_size.
   - If it’s in B2 ghost list, decrease target_t1_size. 
   
4. Blocks eviction (`prepare_store`): This is the “adaptative replacement” part. We get the numer of blocks to evict, and we decide where to evict from:
   - If T1 size is greater than target_t1_size, evict from T1 and therefore add to B1.
   - Else, evict from T2 and therefore add to B2. 
   - Finally, we bound each ghost list size. 

5. Block insertion (`prepare_store`): New blocks are always inserted in T1 and removed from B1/B2 if present there. Then, in step 2, blocks potentially get promoted to T2.

So the interesting thing is the “adaptative” behavior of the core steps, using hits in the ghost lists to adjust the balance: 
- B1 hit means that recent access matters more, and T1 is increased. 
- B2 hit means that frequent access matters more, and T1 is decreased. 

So it kind of self-tunes the recency vs. frrequency trade-off. 

## Test Plan

Added **new** test suite for ARC in `tests/kv_offload/test_cpu_manager.py`.  One can run it by:

```
pytest tests/v1/kv_offload/test_cpu_manager.py -v
```

## Test Result

Test passed:

```
albertomba13 ➜ ~/workspace/vllm$ git:(feature/arc-kv-cache-eviction-policy) pytest tests/v1/kv_offload/test_cpu_manager.py -v

=================================================================================== test session starts ===================================================================================
platform darwin -- Python 3.12.4, pytest-8.3.4, pluggy-1.5.0 -- /Users/aperdomo/.pyenv/versions/3.12.4/bin/python3.12
cachedir: .pytest_cache
rootdir: /Users/aperdomo/workspace/vllm
configfile: pyproject.toml
plugins: cov-6.0.0, anyio-4.9.0, langsmith-0.3.31, dash-3.2.0
collected 10 items                                                                                                                                                                        

tests/v1/kv_offload/test_cpu_manager.py::test_cpu_manager PASSED                                                                                                                    [ 10%]
tests/v1/kv_offload/test_cpu_manager.py::test_arc_manager_basic PASSED                                                                                                              [ 20%]
tests/v1/kv_offload/test_cpu_manager.py::test_arc_manager_t1_to_t2_promotion PASSED                                                                                                 [ 30%]
tests/v1/kv_offload/test_cpu_manager.py::test_arc_manager_eviction_with_load PASSED                                                                                                 [ 40%]
tests/v1/kv_offload/test_cpu_manager.py::test_arc_manager_adaptive_target PASSED                                                                                                    [ 50%]
tests/v1/kv_offload/test_cpu_manager.py::test_arc_manager_t1_t2_eviction_policy PASSED                                                                                              [ 60%]
tests/v1/kv_offload/test_cpu_manager.py::test_arc_manager_ghost_list_bounds PASSED                                                                                                  [ 70%]
tests/v1/kv_offload/test_cpu_manager.py::test_arc_manager_touch_ordering PASSED                                                                                                     [ 80%]
tests/v1/kv_offload/test_cpu_manager.py::test_arc_manager_failed_store PASSED                                                                                                       [ 90%]
tests/v1/kv_offload/test_cpu_manager.py::test_arc_manager_full_scenario PASSED                                                                                                      [100%]

=================================================================================== 10 passed in 1.08s ====================================================================================
```

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

albertoperdomo2 committed:
Implement ARC KV cache eviction policy

Signed-off-by: Alberto Perdomo <aperdomo@redhat.com>

albertoperdomo2 committed:
Get cache capacity from backend

Signed-off-by: Alberto Perdomo <aperdomo@redhat.com>

albertoperdomo2 committed:
Remove incorrect check

Signed-off-by: Alberto Perdomo <aperdomo@redhat.com>

ApostaC reviewed (changes requested):
Did a quick pass. 
Can you explain more about the algorithm here and have more clarity on what `t1/2` and `b1/2` mean in the code? 

I would like to take a look again after having more clarity. Thanks!

albertoperdomo2 committed:
Factory method to create the offloading manager

This change refactors the KV cache offloading system to use a Factory
Pattern for creating eviction policy managers, replacing hardcoded
conditional logic.

Signed-off-by: Alberto Perdomo <aperdomo@redhat.com>

albertoperdomo2 committed:
Fix missing arguments

Signed-off-by: Alberto Perdomo <aperdomo@redhat.com>

albertoperdomo2 committed:
Minor fixes

Signed-off-by: Alberto Perdomo <aperdomo@redhat.com>

albertoperdomo2 commented:
Damn, the suggested fix for the sign-off messed up 😅

ApostaC reviewed (commented):
Overall LGTM. Just a few points:
1. Can you add the description of the algorithm to the top of `arc_manager.py` or the docstring of `ARCOffloadingManager`? This is to make sure other contributors can understand & maintain the code.
2. Can you run some end-to-end performance benchmarking and show under what case this algorithm is better than the default LRU and how much better it is?

albertoperdomo2 committed:
Merge branch 'main' into feature/arc-kv-cache-eviction-policy

albertoperdomo2 committed:
Document algorithm in doc-string

Signed-off-by: Alberto Perdomo <aperdomo@redhat.com>

albertoperdomo2 committed:
Initialize cache capacity properly

Signed-off-by: Alberto Perdomo <aperdomo@redhat.com>

albertoperdomo2 committed:
Minor refactor

Signed-off-by: Alberto Perdomo <aperdomo@redhat.com>

albertoperdomo2 committed:
Merge branch 'main' into feature/arc-kv-cache-eviction-policy

albertoperdomo2 committed:
Fix double definition

Signed-off-by: Alberto Perdomo <aperdomo@redhat.com>

ApostaC reviewed (approved):
LGTM!

albertoperdomo2 committed:
Merge branch 'main' into feature/arc-kv-cache-eviction-policy

orozery commented on vllm/v1/kv_offload/abstract.py:
```diff
@@ -67,6 +71,17 @@ class OffloadingEvent:
 
 class OffloadingManager(ABC):
+    @abstractmethod
+    def __init__(self, backend: "Backend", enable_events: bool = False):
```
This adds a dependency between the `OffloadingManager` and `Backend`.
Not all OffloadingManagers would want this abstraction.
Thus I think we should remove this.

orozery commented on vllm/v1/kv_offload/arc_manager.py:
```diff
@@ -0,0 +1,257 @@
+                            self.backend.free(block)
+
+                            evicted_count += 1
+                            evicted = True
```
```suggestion
```

orozery commented on vllm/v1/kv_offload/arc_manager.py:
```diff
@@ -0,0 +1,257 @@
+
+                if evict_from_t1:
+                    # try to evict the least recently used (oldest) block from T1
+                    evicted = False
```
```suggestion
```

orozery commented on vllm/v1/kv_offload/arc_manager.py:
```diff
@@ -0,0 +1,257 @@
+
+                if not evict_from_t1:
+                    # try to evict the least recently used (oldest) block from T2
+                    evicted = False
```
```suggestion
```

orozery reviewed (changes requested):
Very nice work!
I did a first pass, without diving to the actual logic of ARC.
Thank you!

albertoperdomo2 committed:
Apply suggestions from code review

Co-authored-by: Or Ozeri <or@ozery.com>
Signed-off-by: alberto <aperdomo@redhat.com>

albertoperdomo2 committed:
Fix creating duplicate lists

I think wrapping a dict with a list() was a common pattern when
iterating over a dict and modifying it. However, in these cases we are
just deleting one element and then breaking the loop, so it *should* be
safe.

Signed-off-by: Alberto Perdomo <aperdomo@redhat.com>

albertoperdomo2 commented on vllm/v1/kv_offload/cpu.py:
```diff
@@ -33,16 +33,22 @@ def __init__(self, vllm_config: VllmConfig):
+                block_size=self.offloaded_block_size, num_blocks=self.num_cpu_blocks
+            )
+
+            self._manager = OffloadingManagerFactory.create_manager(
```
@ApostaC what are your thoughts about this?

albertoperdomo2 committed:
Refactor eviction logic

Signed-off-by: Alberto Perdomo <aperdomo@redhat.com>

orozery commented on vllm/v1/kv_offload/arc_manager.py:
```diff
@@ -0,0 +1,248 @@
+        if num_blocks_to_evict > 0:
+            evicted_count = 0
+
+            while evicted_count < num_blocks_to_evict:
```
```suggestion
        while evicted_count < num_blocks_to_evict:
```

+Decrease the indent of all lines in the loop.

orozery reviewed (commented):
Some optimizations to prepare_store + bug fix.
I will go over the other functions tomorrow.

albertoperdomo2 committed:
Update vllm/v1/kv_offload/arc_manager.py

Co-authored-by: Or Ozeri <or@ozery.com>
Signed-off-by: alberto <aperdomo@redhat.com>

albertoperdomo2 committed:
Format

Signed-off-by: Alberto Perdomo <aperdomo@redhat.com>

albertoperdomo2 committed:
Update vllm/v1/kv_offload/arc_manager.py

Co-authored-by: Or Ozeri <or@ozery.com>
Signed-off-by: alberto <aperdomo@redhat.com>

albertoperdomo2 committed:
target_t1_size as a float

Signed-off-by: Alberto Perdomo <aperdomo@redhat.com>

albertoperdomo2 committed:
Promote ghost list hit to MRU position

Signed-off-by: Alberto Perdomo <aperdomo@redhat.com>

albertoperdomo2 committed:
Merge branch 'main' into feature/arc-kv-cache-eviction-policy

orozery commented on vllm/v1/kv_offload/arc_manager.py:
```diff
@@ -0,0 +1,235 @@
+
+        self.cache_capacity: int
+        if hasattr(self.backend, "num_blocks"):
+            self.cache_capacity = self.backend.num_blocks
```
But any `Backend` must implement `get_num_free_blocks`.
Querying `num_blocks` on the other hand seems hacky to me, as it is not defined in the Backend abstract class.

albertoperdomo2 replied:
I agree partially: `get_num_free_blocks` is a fair fallback for this but technically, the actual cache size is determined (or I think it should be) by `num_blocks`. If `num_blocks` is not implemented all the time, I guess we can use `get_num_free_blocks`.

albertoperdomo2 committed:
Update vllm/v1/kv_offload/arc_manager.py

Co-authored-by: Or Ozeri <or@ozery.com>
Signed-off-by: alberto <aperdomo@redhat.com>

albertoperdomo2 committed:
Update vllm/v1/kv_offload/arc_manager.py

Co-authored-by: Or Ozeri <or@ozery.com>
Signed-off-by: alberto <aperdomo@redhat.com>

albertoperdomo2 committed:
Get cache capacity by get_num_free_blocks

Signed-off-by: Alberto Perdomo <aperdomo@redhat.com>

albertoperdomo2 committed:
Remove OffloadingManagerFactory implementation

Signed-off-by: Alberto Perdomo <aperdomo@redhat.com>

albertoperdomo2 committed:
Remove unnecessary code

Signed-off-by: Alberto Perdomo <aperdomo@redhat.com>

albertoperdomo2 committed:
Merge branch 'main' into feature/arc-kv-cache-eviction-policy

orozery reviewed (approved):
LGTM! Thanks!!

albertoperdomo2 committed:
Merge branch 'main' into feature/arc-kv-cache-eviction-policy
